File: app/src/main/java/com/chao/wanmvvm/network/rx/RxSubscriber.java
package com.chao.wanmvvm.network.rx;

import com.chao.wanmvvm.App;
import com.chao.wanmvvm.network.ServerException;
import com.chao.wanmvvm.util.NetworkUtils;
import com.google.gson.JsonParseException;
import io.reactivex.subscribers.DisposableSubscriber;
import org.json.JSONException;
import retrofit2.HttpException;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;

/**
 * @author tqzhang
 */
public abstract class RxSubscriber<T> extends DisposableSubscriber<T> {


    public RxSubscriber() {
        super();
    }

    @Override
    protected void onStart() {
        super.onStart();
        showLoading();
        if (!NetworkUtils.isNetworkAvailable(App.instance())) {
            onNoNetWork();
            cancel();
            return;
        }
    }

    @Override
    public void onComplete() {

    }

    protected void showLoading() {

    }

    protected void onNoNetWork() {

    }

    @Override
    public void onError(Throwable e) {
        String message = null;
        if (e instanceof UnknownHostException) {
            message = "没有网络";
        } else if (e instanceof HttpException) {
            message = "网络错误";
        } else if (e instanceof SocketTimeoutException) {
            message = "网络连接超时";
        } else if (e instanceof JsonParseException
                || e instanceof JSONException) {
            message = "解析错误";
        } else if (e instanceof ConnectException) {
            message = "连接失败";
        } else if (e instanceof ServerException) {
            message = ((ServerException) e).message;
        }
        onFailure(message);
    }

    @Override
    public void onNext(T t) {
        onSuccess(t);
    }

    /**
     * success
     *
     * @param t
     */
    public abstract void onSuccess(T t);

    /**
     * failure
     *
     * @param msg
     */
    public abstract void onFailure(String msg);
}

File: app/src/main/java/com/chao/wanmvvm/network/RxSubscriber.kt
package com.chao.wanmvvm.network

import com.chao.mvvm.util.NetworkUtils
import com.chao.wanmvvm.App
import com.google.gson.JsonParseException
import io.reactivex.subscribers.DisposableSubscriber
import org.json.JSONException
import retrofit2.HttpException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * @author tqzhang
 */
abstract class RxSubscriber<T> : DisposableSubscriber<T>() {

    override fun onStart() {
        super.onStart()
        showLoading()
        if (!NetworkUtils.isNetworkAvailable(App.instance())) {
            onNoNetWork()
            cancel()
            return
        }
    }

    override fun onComplete() {

    }

    fun showLoading() {

    }

    fun onNoNetWork() {

    }

    override fun onError(e: Throwable) {
        var message: String? = null
        if (e is UnknownHostException) {
            message = "没有网络"
        } else if (e is HttpException) {
            message = "网络错误"
        } else if (e is SocketTimeoutException) {
            message = "网络连接超时"
        } else if (e is JsonParseException || e is JSONException) {
            message = "解析错误"
        } else if (e is ConnectException) {
            message = "连接失败"
        } else if (e is ServerException) {
            message = e.message
        }
        onFailure(message)
    }

    override fun onNext(t: T) {
        onSuccess(t)
    }

    /**
     * success
     *
     * @param t
     */
    abstract fun onSuccess(t: T)

    /**
     * failure
     *
     * @param msg
     */
    abstract fun onFailure(msg: String?)
}
